rfcs/experimental/flow_graph_wait_try_put_and_wait/README.md

```diff
@@ -0,0 +1,608 @@
+since it was submitted after the blue tasks.
+
+``f1`` and ``f2`` are ``unlimited`` ``function_node``s, so the tasks ordering is arbitrary since tasks for each computed item are spawned.
+
+``join`` is a queueing ``join_node`` and it should preserve the ordering of items as they are processed by ``f1`` and ``f2``. The tasks are shown
+as partially blue and partially red since the output of red task from ``f1`` can be combined with the output of blue task from ``f2`` and wise versa.
+
+Since ``post_process`` is a serial ``function_node``, the ordering of the tasks will be the same as in ``join``.
+
+The ``try_put_and_wait`` is expected to exit when all of the red (including partially red) tasks are completed. It may require some amount of blue tasks
+to be completed as well, e.g. to execute previously stored tasks in the ``function_node``'s queue. 
+
+Completion of all blue tasks as in ``wait_for_all`` is not guaranteed.
+
+## Feature Design
+
+This feature is implemented by creating an instance of special private class ``message_metainfo`` for each input message in ``try_put_and_wait``. This instance
+wraps the pointer to the dedicated ``wait_context`` object representing the reference counter for single input message.
+
+``message_metainfo`` is broadcasted through the graph together with the message itself. The actual value and the type of the message can be changed during the computation, but the stored 
+metainformation is preserved.
+
+The reference counter in the ``wait_context`` inside of the ``message_metainfo`` is increased in the following cases:
+* The task associated with the computations of the ``try_put_and_wait`` input message is created (*).
+* The item associated with the computations of the desired input is stored in the buffering node or in the internal buffer of the other node type.
+* When the ``continue_msg`` with the non-empty associated waiter is received by ``continue_node``. In this case, the node buffers the ``message_metainfo``s received from each
+  predecessor and increases the underlying reference counter to prolonge the wait until signals from each predecessor is received. 
+
+The reference counter is decreased when:
+* The task associated with the computations of the ``try_put_and_wait`` input is finalized. If the output of the task should be propagated to the successors of the node, it is done
+  from the task body and hence the reference counter is decreased only after creating the task for each successor and increasing the reference counter for them.
+* The item associated with the computations is taken from the buffering node or from the internal buffer.
+* When the desired number of signals from the predecessors was received by the ``continue_node``. This case is equivalent to retrieving the set of buffered ``message_metainfo``s received previously
+  from each predecessor. 
```

Comment:
For these cases, how it is ensured that the reference counters are not decreased prematurely?

Reply:
`continue_node` creates a copy of the stored metainfo merged from each of the predecessors and creates a task for executing the body that holds an additional reference on the counter (similarly to any task with the stored metainfo). 
And only after spawning the task, the reference counters on the copy are decreased.

I have added a note that the reference counter is decreased after spawning a task.